[run-benchmark] Add motionmark1.3 plan variant that enforces 15fps for low-end devices

clopez opened the pull request:
#### 9a083c67ec38fa2f7e72bf42745b083b5b7db207
<pre>
[run-benchmark] Add motionmark1.3 plan variant that enforces 15fps for low-end devices
<a href="https://bugs.webkit.org/show_bug.cgi?id=276127">https://bugs.webkit.org/show_bug.cgi?id=276127</a>

Reviewed by Nikolas Zimmermann.

With the default MotionMark settings on low-end devices like the RPi4 the results
are so low (1-3 pts) that it is impossible to use this benchmark to performance
regression tracking.

This adds a variant of motionmark-1.3 with a target frame-rate of 15FPS, that
way on the RPi4 boards we can get scores around 100-600 which are much more useful
to track performance regressions.
We tested with 30FPS but that is still too much for this boards with MotionMark.

The animations on the screen also look better, you can clear see more complexity
in the animations even when you can see that that it animates with low FPS.

In order to avoid copying-pasting the whole plan itself this patch also introduces
a way of importing another plan inside a run-benchmark .plan file via the key
`import_plan_file` and also allows to apply an extra patch via the key `extra_patch`
that will be applied at the end of the patch process.

That allows to see more clearly what this plan changes on top of the default MotionMark
plan.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_builder.py:
(BenchmarkBuilder.__enter__):
* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner._load_plan_data):
* Tools/Scripts/webkitpy/benchmark_runner/data/patches/extra/MotionMark1.3-15FPS.patch: Added.
* Tools/Scripts/webkitpy/benchmark_runner/data/plans/motionmark1.3-15fps.plan: Added.
* Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
(BrowserPerfDashRunner._get_test_version_string): Include the imported plan into the hash-version calculation-

Canonical link: <a href="https://commits.webkit.org/280660@main">https://commits.webkit.org/280660@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ebd23047545d8fea4833d0317ecaa791ebd91e48

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57208 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9683 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60829 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7651 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59336 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44160 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7841 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46317 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5384 "Passed tests") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59238 "Build is in progress. Recent messages:OS: Unknown (10.15.7), Xcode: 11.7; Running apply-patch; Checked out pull request; Passed webkitperl tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34285 "Found 60 new test failures: accessibility/accessibility-node-reparent.html accessibility/ancestor-computation.html accessibility/announcement-notification.html accessibility/aria-braillelabel.html accessibility/aria-brailleroledescription.html accessibility/aria-checked-mixed-value.html accessibility/aria-current-state-changed-notification.html accessibility/aria-current.html accessibility/aria-describedby-on-input.html accessibility/aria-description.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49391 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27177 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/56734 "Passed tests") | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31067 "Found 60 new test failures: imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html imported/w3c/web-platform-tests/FileAPI/FileReader/Progress_event_bubbles_cancelable.html imported/w3c/web-platform-tests/FileAPI/FileReader/workers.html imported/w3c/web-platform-tests/FileAPI/FileReaderSync.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-dom.window.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-endings.html ... (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6707 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6656 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6977 "Found 1 new test failure: media/video-transformed.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62509 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1121 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7075 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53578 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1126 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49430 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53651 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/941 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8537 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32365 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33450 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34535 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33196 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann reviewed (approved):
Looks much nicer - @shallawa  your suggestion was implemented to avoid the plan copy.

dewei-zhu commented on Tools/Scripts/webkitpy/benchmark_runner/data/plans/motionmark1.3-15fps.plan:
```diff
@@ -0,0 +1,5 @@
+{
+    "import_plan_file": "motionmark1.3.plan",
+    "output_file": "motionmark1.3-15fps.result",
+    "extra_patch": "data/patches/extra/MotionMark1.3-15FPS.patch"
```
Since we are adding a new field for more patches, let's allow this variable to provide a list of patches instead of just one. something like `"extra_patches": [ "data/patches/extra/MotionMark1.3-15FPS.patch"]` for better future compatibility.

clopez replied:
Good idea! will implement that :)

dewei-zhu commented on Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
```diff
@@ -166,6 +166,14 @@ def _get_test_version_string(self, plan_name):
             version_hash.update(plan_content)
             plan_dict = json.loads(plan_content)
+            if 'import_plan_file' in plan_dict:
+                imported_plan_file = os.path.join(self._plandir, plan_dict['import_plan_file'])
```
Maybe use `plan_dict.pop('import_plan_file')` so that the remaining key value pairs on `plan_dict` are pure override.

clopez replied:
yes, good point

dewei-zhu commented on Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
```diff
@@ -166,6 +166,14 @@ def _get_test_version_string(self, plan_name):
+                imported_plan_dict.update(plan_dict)
+                plan_dict = imported_plan_dict
             for plan_key in plan_dict:
                 if plan_key.endswith('_patch'):
```
Should we also do similar thing for the new `extra_patches` field?
Also, the way `version_hash` is calculated doesn't seem to be stable, and even possible that even the same plan file with shuffled key order would result in different hash, not sure if that is intended.

clopez replied:
yes, content of patches should be added.
And you are right, than shuffling the keys in the plan file would change the hash but that is ok, I guess. The idea of this hash_version is to detect when something changes on the .plan files or any of their patches, so when you compare two runs for a given plan if you see two different hash_versions then that is an indication that something changed on the test definition and that is worth investigating what the change was about (it may explain or not the differences on the results between the two runs)

clopez committed:
[run-benchmark] Add motionmark1.3 plan variant that enforces 15fps for low-end devices
https://bugs.webkit.org/show_bug.cgi?id=276127

Reviewed by Nikolas Zimmermann.

With the default MotionMark settings on low-end devices like the RPi4 the results
are so low (1-3 pts) that it is impossible to use this benchmark to performance
regression tracking.

This adds a variant of motionmark-1.3 with a target frame-rate of 15FPS, that
way on the RPi4 boards we can get scores around 100-600 which are much more useful
to track performance regressions.
We tested with 30FPS but that is still too much for this boards with MotionMark.

The animations on the screen also look better, you can clear see more complexity
in the animations even when you can see that that it animates with low FPS.

In order to avoid copying-pasting the whole plan itself this patch also introduces
a way of importing another plan inside a run-benchmark .plan file via the key
`import_plan_file` and also allows to apply an extra patch via the key `extra_patch`
that will be applied at the end of the patch process.

That allows to see more clearly what this plan changes on top of the default MotionMark
plan.

* Tools/Scripts/webkitpy/benchmark_runner/benchmark_builder.py:
(BenchmarkBuilder.__enter__):
* Tools/Scripts/webkitpy/benchmark_runner/benchmark_runner.py:
(BenchmarkRunner._load_plan_data):
* Tools/Scripts/webkitpy/benchmark_runner/data/patches/extra/MotionMark1.3-15FPS.patch: Added.
* Tools/Scripts/webkitpy/benchmark_runner/data/plans/motionmark1.3-15fps.plan: Added.
* Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
(BrowserPerfDashRunner._get_test_version_string): Include the imported plan into the hash-version calculation-

Canonical link: https://commits.webkit.org/280660@main
